File: src/main/java/com/regenerationforrged/data/worldgen/preset/settings/WorldSettings.java
package com.regenerationforrged.data.worldgen.preset.settings;

import com.mojang.serialization.Codec;
import com.mojang.serialization.codecs.RecordCodecBuilder;

import com.regenerationforrged.world.worldgen.cell.continent.MushroomIslandPopulator;
import com.regenerationforrged.world.worldgen.noise.function.DistanceFunction;

public class WorldSettings {
	public static final Codec<WorldSettings> CODEC = RecordCodecBuilder.create(instance -> instance.group(
		Continent.CODEC.fieldOf("continent").forGetter((o) -> o.continent),
		ControlPoints.CODEC.fieldOf("controlPoints").forGetter((o) -> o.controlPoints),
		Properties.CODEC.fieldOf("properties").forGetter((o) -> o.properties)
	).apply(instance, WorldSettings::new));
	
    public Continent continent;
    public ControlPoints controlPoints;
    public Properties properties;
    
    public WorldSettings(Continent continent, ControlPoints controlPoints, Properties properties) {
        this.continent = continent;
        this.controlPoints = controlPoints;
        this.properties = properties;
    }
    
    public WorldSettings copy() {
    	return new WorldSettings(this.continent.copy(), this.controlPoints.copy(), this.properties.copy());
    }
    
    public static class Continent {
    	public static final Codec<Continent> CODEC = RecordCodecBuilder.create(instance -> instance.group(
    		ContinentType.CODEC.fieldOf("continentType").forGetter((o) -> o.continentType),
    		DistanceFunction.CODEC.optionalFieldOf("continentShape", DistanceFunction.EUCLIDEAN).forGetter((o) -> o.continentShape),
    		Codec.INT.fieldOf("continentScale").forGetter((o) -> o.continentScale),
    		Codec.FLOAT.fieldOf("continentJitter").forGetter((o) -> o.continentJitter),
    		Codec.FLOAT.optionalFieldOf("continentSkipping", 0.25F).forGetter((o) -> o.continentSkipping),
    		Codec.FLOAT.optionalFieldOf("continentSizeVariance", 0.25F).forGetter((o) -> o.continentSizeVariance),
    		Codec.INT.optionalFieldOf("continentNoiseOctaves", 5).forGetter((o) -> o.continentNoiseOctaves),
    		Codec.FLOAT.optionalFieldOf("continentNoiseGain", 0.26F).forGetter((o) -> o.continentNoiseGain),
    		Codec.FLOAT.optionalFieldOf("continentNoiseLacunarity", 4.33F).forGetter((o) -> o.continentNoiseLacunarity),
			Codec.FLOAT.optionalFieldOf("continentWarpStrength", 25F).forGetter((o) -> o.continentsWarpStrength),
			Codec.FLOAT.optionalFieldOf("continentWarpScale" 200F).forGetter((o) -> o.continentWarpScale),
			Codec.FLOAT.optionalFieldOf("coastSharpness" 0.5F).forGetter((o) -> o.coastSharpness),
			Codec.FLOAT.optionalFieldOf("valleyErosion" 0.5F).forGetter((o) -> o.valleyErosion),
			Codec.FLOAT.optionalFieldOf("valleyWeirdness" 1F).forGetter((o) -> o.valleyWeirdness),
			Codec.FLOAT.optionalFieldOf("riverErosion" 1.5F).forGetter((o) -> o.riverErosion),
			Codec.FLOAT.optionalFieldOf("riverWeirdness" 1.5F).forGetter((o) -> o.riverWeirdness),
			Codec.FLOAT.optionalFieldOf("lakeWeirdNess" 0.5F).forGetter((o) -> o.lakeWeirdness),
			Codec.INT.optionalFieldOf("beachNoiseScale").forGetter((o) -> o.beachNoiseScale),
			Codec.INT.optionalFieldOf("beachHeight").forGetter((o) -> o.beachHeight)
    	).apply(instance, Continent::new));
    	
        public ContinentType continentType;
        public DistanceFunction continentShape;
        public int continentScale;
        public float continentJitter;
        public float continentSkipping;
        public float continentSizeVariance;
        public int continentNoiseOctaves;
        public float continentNoiseGain;
        public float continentNoiseLacunarity;
		public float continentWarpStrength;
		public float contientWarpScale;
		public float coastSharpness;
		public float valleyErosion;
		public float valleyWeirdness;
		public float riverErosion;
		public float riverWeirdness;
		public float lakeWeirdness;
		public int beachNoiseScale;
		public int beachHeight;
        
        public Continent(ContinentType continentType, DistanceFunction continentShape, int continentScale, float continentJitter, float continentSkipping, float continentSizeVariance, int continentNoiseOctaves, float continentNoiseGain, float continentNoiseLacunarity, float continentWarpStrength, float continentWarpScale, float coastSharpness,
						valleyErosion valleyErosion, valleyWeirdness valleyWeirdness, riverErosion riverErosion, riverWeirdness riverWeirdness, lakeWeirdness,
						beachNoiseScale beachNoiseScale, beachHeight beachHeight) {
            this.continentType = continentType;
            this.continentShape = continentShape;
            this.continentScale = continentScale;
            this.continentJitter = continentJitter;
            this.continentSkipping = continentSkipping;
            this.continentSizeVariance = continentSizeVariance;
            this.continentNoiseOctaves = continentNoiseOctaves;
            this.continentNoiseGain = continentNoiseGain;
            this.continentNoiseLacunarity = continentNoiseLacunarity;
			this.continentWarpStrength = continentWarpStrength;
			this.continentWarpScale = continentWarpScale;
			this.coastSharpness = coastSharpness;
			this.valleyErosion = valleyErosion;
			this.valleyWeirdness = valleyWeirdness;
			this.riverErosion = riverErosion;
			this.riverWeirdness = riverWeirdness;
			this.lakeWeirdness = lakeWeirdness;
			this.beachNoiseScale = beachNoiseScale;
			this.beachHeight = beachHeight;
        }
        
        public Continent copy() {
        	return new Continent(this.continentType, this.continentShape, this.continentScale, this.continentJitter, this.continentSkipping, this.continentSizeVariance, this.continentNoiseOctaves, this.continentNoiseGain, this.continentNoiseLacunarity, this.continentWarpStrength, this.continentWarpScale, this.coastSharpness,
								this.valleyErosion, this.valleyWeirdness, this.riverErosion, this.riverWeirdness,
								this.lakeWeirdness, this.beachNoiseScale, this.beachHeight);
        }
    }
    
    public static class ControlPoints {
    	public static final Codec<ControlPoints> CODEC = RecordCodecBuilder.create(instance -> instance.group(
        	Codec.FLOAT.optionalFieldOf("mushroomFieldsInland", MushroomIslandPopulator.DEFAULT_INLAND_POINT).forGetter((o) -> o.mushroomFieldsInland),
        	Codec.FLOAT.optionalFieldOf("mushroomFieldsCoast", MushroomIslandPopulator.DEFAULT_COAST_POINT).forGetter((o) -> o.mushroomFieldsCoast),
    		Codec.FLOAT.fieldOf("deepOcean").forGetter((o) -> o.deepOcean),
    		Codec.FLOAT.fieldOf("shallowOcean").forGetter((o) -> o.shallowOcean),
    		Codec.FLOAT.fieldOf("beach").forGetter((o) -> o.beach),
    		Codec.FLOAT.fieldOf("coast").forGetter((o) -> o.coast),
    		Codec.FLOAT.fieldOf("inland").forGetter((o) -> o.inland),
			Codec.FLOAT.optionalFieldOf("coastLineBlend" 0.25F).forGetter((o) -> o.coastLineBlend)
        ).apply(instance, ControlPoints::new));

    	public float mushroomFieldsInland;
    	public float mushroomFieldsCoast;
        public float deepOcean;
        public float shallowOcean;
        public float beach;
        public float coast;
        public float inland;
		public float coastLineBlend
        
        public ControlPoints(float mushroomFieldsInland, float mushroomFieldsCoast, float deepOcean, float shallowOcean, float beach, float coast, float inland, float coastLineBlend) {
        	this.mushroomFieldsInland = mushroomFieldsInland;
        	this.mushroomFieldsCoast = mushroomFieldsCoast;
            this.deepOcean = deepOcean;
            this.shallowOcean = shallowOcean;
            this.beach = beach;
            this.coast = coast;
            this.inland = inland;
			this.coastLineBlend = coastLineBlend;
        }
        
        public ControlPoints copy() {
        	return new ControlPoints(this.mushroomFieldsInland, this.mushroomFieldsCoast, this.deepOcean, this.shallowOcean, this.beach, this.coast, this.inland, this.coatsLineBlend);
        }
    }
    
    public static class Properties {
    	public static final Codec<Properties> CODEC = RecordCodecBuilder.create(instance -> instance.group(
    		SpawnType.CODEC.fieldOf("spawnType").forGetter((o) -> o.spawnType),
    		Codec.INT.fieldOf("worldHeight").forGetter((o) -> o.worldHeight),
			Codec.INT.fieldOf("worldMinY").forGetter((o) -> o.worldMinY),
			Codec.FLOAT.fieldOf("riverScale").forGetter((o) -> o.riverScale),
			Codec.INT.fieldOf("structureSeparation").forGetter((o) -> o.structureSepararion),
    		Codec.INT.optionalFieldOf("worldDepth", -64).forGetter((o) -> o.worldDepth),
    		Codec.INT.fieldOf("seaLevel").forGetter((o) -> o.seaLevel),
    		Codec.INT.optionalFieldOf("lavaLevel", -54).forGetter((o) -> o.lavaLevel),
			Codec.BOOL.optionalFieldOf("JavaEngineWorld", true).forGetter((o) -> o.JavaEngineWorld)
    	).apply(instance, Properties::new));
    	
        public SpawnType spawnType;
        public int worldHeight;
		public int worldMinY;
		public float riverScale;
		public int structureSeparation;
        public int worldDepth;
        public int seaLevel;
        public int lavaLevel;
		public boolean JavaEngineWorld;
        
        public Properties(SpawnType spawnType, int worldHeight,
						  int worldMinY, float riverScale, int structureSeparation,
                        int worldDepth, int seaLevel, int lavaLevel, bool JavaEngineWorld) {
        	this.spawnType = spawnType;
        	this.worldHeight = worldHeight;
			this.worldMinY = worldMinY;
			this.riverScale = riverScale;
			this.structureSeparation = structureSeparation;
        	this.worldDepth = worldDepth;
        	this.seaLevel = seaLevel;
        	this.lavaLevel = lavaLevel;
			this.JavaEngineWorld = JavaEngineWorld;
        }
        
        public Properties copy() {
        	return new Properties(this.spawnType, this.worldHeight, 
								  this.worldMinY, this.riverScale, this.structureSeparation, this.worldDepth, this.seaLevel, this.lavaLevel, this.JavaEngineWorld);
        }
        
        @Deprecated
        public int terrainScaler() {
        	return Math.min(this.worldHeight, 640);
        }
    }
}

File: src2/main/kotlin/com/mbz/data/worldgen/preset/settings/WorldSettings.kt
package com.mbz.data.worldgen.preset.settings

import com.mbz.world.worldgen.cell.continent.MushroomIslandPopulator
import com.mbz.world.worldgen.noise.function.DistanceFunction
import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder

class WorldSettings(
    var continent: Continent,
    var controlPoints: ControlPoints,
    var properties: Properties
) {
    fun copy() = WorldSettings(continent.copy(), controlPoints.copy(), properties.copy())

    // --- Cấu hình Lục địa ---
    class Continent(
        var continentType: ContinentType,
        var continentShape: DistanceFunction,
        var continentScale: Int,
        var continentJitter: Float,
        var continentSkipping: Float,
        var continentSizeVariance: Float,
        var continentNoiseOctaves: Int,
        var continentNoiseGain: Float,
        var continentNoiseLacunarity: Float,
        var continentWarpStrength: Float,
        var continentWarpScale: Float
    ) {
        fun copy() = Continent(
            continentType, continentShape, continentScale, continentJitter, continentSkipping,
            continentSizeVariance, continentNoiseOctaves, continentNoiseGain, continentNoiseLacunarity,
            continentWarpStrength, continentWarpScale
        )

        companion object {
            @JvmField
            val CODEC: Codec<Continent> = RecordCodecBuilder.create { i ->
                i.group(
                    ContinentType.CODEC.fieldOf("continentType").forGetter { it.continentType },
                    DistanceFunction.CODEC.optionalFieldOf("continentShape", DistanceFunction.EUCLIDEAN).forGetter { it.continentShape },
                    Codec.INT.fieldOf("continentScale").forGetter { it.continentScale },
                    Codec.FLOAT.fieldOf("continentJitter").forGetter { it.continentJitter },
                    Codec.FLOAT.optionalFieldOf("continentSkipping", 0.25f).forGetter { it.continentSkipping },
                    Codec.FLOAT.optionalFieldOf("continentSizeVariance", 0.25f).forGetter { it.continentSizeVariance },
                    Codec.INT.optionalFieldOf("continentNoiseOctaves", 5).forGetter { it.continentNoiseOctaves },
                    Codec.FLOAT.optionalFieldOf("continentNoiseGain", 0.26f).forGetter { it.continentNoiseGain },
                    Codec.FLOAT.optionalFieldOf("continentNoiseLacunarity", 4.33f).forGetter { it.continentNoiseLacunarity },
                    Codec.FLOAT.optionalFieldOf("continentWarpStrength", 25f).forGetter { it.continentWarpStrength },
                    Codec.FLOAT.optionalFieldOf("continentWarpScale", 200f).forGetter { it.continentWarpScale }
                ).apply(i, ::Continent)
            }
        }
    }

    // --- Các điểm kiểm soát độ cao (Ocean, Beach, Coast, Inland) ---
    class ControlPoints(
        var mushroomFieldsInland: Float,
        var mushroomFieldsCoast: Float,
        var deepOcean: Float,
        var shallowOcean: Float,
        var beach: Float,
        var coast: Float,
        var inland: Float,
        var coastLineBlend: Float
    ) {
        fun copy() = ControlPoints(
            mushroomFieldsInland, mushroomFieldsCoast, deepOcean, shallowOcean, 
            beach, coast, inland, coastLineBlend
        )

        companion object {
            @JvmField
            val CODEC: Codec<ControlPoints> = RecordCodecBuilder.create { i ->
                i.group(
                    Codec.FLOAT.optionalFieldOf("mushroomFieldsInland", MushroomIslandPopulator.DEFAULT_INLAND_POINT).forGetter { it.mushroomFieldsInland },
                    Codec.FLOAT.optionalFieldOf("mushroomFieldsCoast", MushroomIslandPopulator.DEFAULT_COAST_POINT).forGetter { it.mushroomFieldsCoast },
                    Codec.FLOAT.fieldOf("deepOcean").forGetter { it.deepOcean },
                    Codec.FLOAT.fieldOf("shallowOcean").forGetter { it.shallowOcean },
                    Codec.FLOAT.fieldOf("beach").forGetter { it.beach },
                    Codec.FLOAT.fieldOf("coast").forGetter { it.coast },
                    Codec.FLOAT.fieldOf("inland").forGetter { it.inland },
                    Codec.FLOAT.optionalFieldOf("coastLineBlend", 0.25f).forGetter { it.coastLineBlend }
                ).apply(i, ::ControlPoints)
            }
        }
    }

    // --- Thuộc tính vật lý của thế giới ---
    class Properties(
        var spawnType: SpawnType,
        var worldHeight: Int,
        var worldMinY: Int,
        var riverScale: Float,
        var structureSeparation: Int,
        var worldDepth: Int,
        var seaLevel: Int,
        var lavaLevel: Int,
        var javaEngineWorld: Boolean
    ) {
        fun copy() = Properties(
            spawnType, worldHeight, worldMinY, riverScale, structureSeparation, 
            worldDepth, seaLevel, lavaLevel, javaEngineWorld
        )

        @Deprecated("Sử dụng worldHeight trực tiếp")
        fun terrainScaler(): Int = Math.min(this.worldHeight, 640)

        companion object {
            @JvmField
            val CODEC: Codec<Properties> = RecordCodecBuilder.create { i ->
                i.group(
                    SpawnType.CODEC.fieldOf("spawnType").forGetter { it.spawnType },
                    Codec.INT.fieldOf("worldHeight").forGetter { it.worldHeight },
                    Codec.INT.fieldOf("worldMinY").forGetter { it.worldMinY },
                    Codec.FLOAT.fieldOf("riverScale").forGetter { it.riverScale },
                    Codec.INT.fieldOf("structureSeparation").forGetter { it.structureSeparation },
                    Codec.INT.optionalFieldOf("worldDepth", -64).forGetter { it.worldDepth },
                    Codec.INT.fieldOf("seaLevel").forGetter { it.seaLevel },
                    Codec.INT.optionalFieldOf("lavaLevel", -54).forGetter { it.lavaLevel },
                    Codec.BOOL.optionalFieldOf("JavaEngineWorld", true).forGetter { it.javaEngineWorld }
                ).apply(i, ::Properties)
            }
        }
    }

    companion object {
        @JvmField
        val CODEC: Codec<WorldSettings> = RecordCodecBuilder.create { i ->
            i.group(
                Continent.CODEC.fieldOf("continent").forGetter { it.continent },
                ControlPoints.CODEC.fieldOf("controlPoints").forGetter { it.controlPoints },
                Properties.CODEC.fieldOf("properties").forGetter { it.properties }
            ).apply(i, ::WorldSettings)
        }
    }
}
